### PR TITLE
xds/federation: e2e tests

### PR DESCRIPTION
**Pull request by menghanl**
Also moved func `BuildResourceName` to testutils to be shared

RELEASE NOTES: N/A

### CODE REVIEW

File: xds/internal/testutils/testutils.go [hunk 1]
```diff
@@ -17,3 +17,29 @@
 
 // Package testutils provides utility types, for use in xds tests.
 package testutils
+
+import (
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource/version"
+)
+
+func BuildResourceName(typ xdsresource.ResourceType, auth, id string, ctxParams map[string]string) string {
```
Comment (easwars) [inline xds/internal/testutils/testutils.go:26]: Needs a comment to make vet happy.

File: xds/internal/test/xds_client_federation_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+// - LDS: old style, no authority (default authority)
+// - RDS: new style, in a different authority
+// - CDS: old style, no authority (default authority)
+// - EDS: new style, in a different authority
+func (s) TestClientSideFederation(t *testing.T) {
+	defer func() func() {
+		old := envconfig.XDSFederation
+		envconfig.XDSFederation = true
+		return func() { envconfig.XDSFederation = old }
+	}()()
```
Comment (easwars) [inline xds/internal/test/xds_client_federation_test.go:54]: This double `func` is super hard to read. Why do we need to do it this way?

File: xds/internal/test/xds_client_federation_test.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+	})
+	if err != nil {
+		t.Fatalf("Failed to create bootstrap file: %v", err)
+	}
+
+	resolver, err := xds.NewXDSResolverWithConfigForTesting(bootstrapContents)
+	if err != nil {
+		t.Fatalf("Failed to create xDS resolver for testing: %v", err)
+	}
+	port, cleanup2 := clientSetup(t, &testService{})
```
Comment (easwars) [inline xds/internal/test/xds_client_federation_test.go:89]: Nit: there is no var `cleanup` for this to be `cleanup2`

File: xds/internal/test/xds_client_federation_test.go (hunk 2 above)
Comment (menghanl) [reply xds/internal/test/xds_client_federation_test.go]: Deleted

File: xds/internal/test/xds_client_federation_test.go (hunk 3 above)
Comment (menghanl) [reply xds/internal/test/xds_client_federation_test.go]: Done

File: xds/internal/testutils/testutils.go (hunk 1 above)
Comment (menghanl) [reply xds/internal/testutils/testutils.go:28]: Done

**Review by easwars — APPROVED** [review]

**Pull request merged**